### PR TITLE
Extend Native-ERC20 limits functionality on contracts v1.1

### PR DESCRIPTION
**Pull request by patitonar**
Relates to issue #108

Gist for deployed HomeBridge: https://gist.github.com/patitonar/df585134034f5b6fbe01928fc6e00e36
Gist for deployed ForeignBridge: https://gist.github.com/patitonar/0ff940d070ef5c4d3dbd70e44eeae361

### CODE REVIEW

File: contracts/upgradeable_contracts/U_ForeignBridge.sol [hunk 1]
```diff
@@ -125,6 +143,7 @@ contract ForeignBridge is ERC677Receiver, BasicBridge {
     function deposit(address recipient, uint256 value, bytes32 transactionHash) external onlyValidator {
         bytes32 hashMsg = keccak256(recipient, value, transactionHash);
         bytes32 hashSender = keccak256(msg.sender, hashMsg);
+        require(withinHomeLimit(value));
```
Comment (akolotov) [inline contracts/upgradeable_contracts/U_ForeignBridge.sol]: It is necessary to check limits and increment the daily limit by the value to check the limit in the next time properly. Please refer to the code in the `onTokenTransfer` method.

File: contracts/upgradeable_contracts/U_ForeignBridge.sol [hunk 2]
```diff
@@ -215,11 +234,20 @@ contract ForeignBridge is ERC677Receiver, BasicBridge {
         emit DailyLimit(_foreignDailyLimit);
     }
 
+    function setHomeDailyLimit(uint256 _homeDailyLimit) external onlyOwner {
+        uintStorage[keccak256("homeDailyLimit")] = _homeDailyLimit;
```
Comment (akolotov) [inline contracts/upgradeable_contracts/U_ForeignBridge.sol]: Does it make sense to emit an event similarly it is done in `setForeignDailyLimit`?

File: contracts/upgradeable_contracts/U_ForeignBridge.sol [hunk 3]
```diff
@@ -215,11 +234,20 @@ contract ForeignBridge is ERC677Receiver, BasicBridge {
...
+    }
+
     function withinLimit(uint256 _amount) public view returns(bool) {
         uint256 nextLimit = totalSpentPerDay(getCurrentDay()).add(_amount);
         return foreignDailyLimit() >= nextLimit && _amount <= maxPerTx() && _amount >= minPerTx();
     }
 
+    function withinHomeLimit(uint256 _amount) public view returns(bool) {
+        uint256 nextLimit = totalSpentPerDay(getCurrentDay()).add(_amount);
```
Comment (akolotov) [inline contracts/upgradeable_contracts/U_ForeignBridge.sol]: the variable `totalSpentPerDay` is logically linked with the daily limit related to the operations on the foreign network. So, this check is not correct. Please introduce another structure to track the daily limit related to operations on the home network.

File: contracts/upgradeable_contracts/U_ForeignBridge.sol (hunk 2 above)
Comment (patitonar) [reply contracts/upgradeable_contracts/U_ForeignBridge.sol]: Yes, added an event.

File: contracts/upgradeable_contracts/U_ForeignBridge.sol (hunk 3 above)
Comment (patitonar) [reply contracts/upgradeable_contracts/U_ForeignBridge.sol]: Thanks, Fixed.

File: contracts/upgradeable_contracts/U_ForeignBridge.sol (hunk 1 above)
Comment (patitonar) [reply contracts/upgradeable_contracts/U_ForeignBridge.sol]: Thanks, Done.

**Comment by patitonar** [comment]
@akolotov Followed the steps you mentioned. Everything worked as expected. Here are some notes from my actions:

1. Updated gist with last changes:
Gist for deployed HomeBridge: https://gist.github.com/patitonar/df585134034f5b6fbe01928fc6e00e36
Gist for deployed ForeignBridge: https://gist.github.com/patitonar/0ff940d070ef5c4d3dbd70e44eeae361

2. Deployed bridge contracts using deploy script and run rust bridge on `Sokol - Kovan`. 
- `v1.0 poa-bridge-contracts` 
- `v0.3.0 poa-bridge`

```
    [   Home  ] HomeBridge: 0xDf1DA79e7908702bBa07860Ff4961EE5bFE69aAE at block 5809810
    [ Foreign ] ForeignBridge: 0x3518297716D5FB68e625Abcb4C491B03D36711ed at block 9499616
    [ Foreign ] POA20: 0x0D820bb285038a724efE599671CD40e134C192F0
```

3. Used `Bridge UI v1.0.0` to send transactions. After some tx the state was:
```
    Home
    Remaining Daily POA Quota: 29,999,998.9 POA
    Locked POA in Bridge Contract: 1.00 POA

    Foreign
    Remaining Daily POA20 Quota: 14,999,999.9 POA20
    POA20 Tokens Amount: 1.00 POA20
```

4. Compiled contracts from gist files
```
        npm install solc@0.4.24 
        solcjs --version  // 0.4.24+commit.e67f0147.Emscripten.clang
        ../node_modules/solc/solcjs gist_home.sol --bin 
        ../node_modules/solc/solcjs gist_foreign.sol --bin
        ../node_modules/solc/solcjs gist_home.sol --abi 
        ../node_modules/solc/solcjs gist_foreign.sol --abi
```
Used [eth-cli](https://github.com/protofire/eth-cli) tool for the following steps.
- Deploy new bridge contract on Home
```
eth deploy <PK> gist_home_sol_HomeBridge.bin --sokol 
// output: 0xee2Cf1d8d6236AF0Fd100D196Ffbc49F8fe4D1d4
```
- Loaded Home contracts
```
eth load-contract HomeBridge.abi  0xDf1DA79e7908702bBa07860Ff4961EE5bFE69aAE EternalStorageProxy.json  0xDf1DA79e7908702bBa07860Ff4961EE5bFE69aAE --sokol
``` 
then
```
web3.eth.accounts.wallet.add(<0xPK>) // to unlock account
EternalStorageProxy.methods.version().call() // '1'
EternalStorageProxy.methods.upgradeTo('2', '0xee2Cf1d8d6236AF0Fd100D196Ffbc49F8fe4D1d4').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
EternalStorageProxy.methods.version().call() // '2'
```
- Deployed new bridge contract on Foreign
```
eth deploy <PK> gist_foreign_sol_ForeignBridge.bin --kovan
  // output: 0x2A9782f7fF77101C59Fa3c7AadCd61E87C933Bc8
```
- Loaded Foreign contracts
```
eth load-contract ForeignBridge.abi 0x3518297716D5FB68e625Abcb4C491B03D36711ed EternalStorageProxy.json  0x3518297716D5FB68e625Abcb4C491B03D36711ed --kovan
``` 
then
```
web3.eth.accounts.wallet.add(<0xPK>) // to unlock account
EternalStorageProxy.methods.version().call() // '1'
EternalStorageProxy.methods.upgradeTo('2', '0x2A9782f7fF77101C59Fa3c7AadCd61E87C933Bc8').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
EternalStorageProxy.methods.version().call() // '2'
```

5.
- Loaded contracts on Home and then:
```
    HomeBridge.methods.setForeignDailyLimit('5000000000000000000').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
    HomeBridge.methods.setForeignMaxPerTx('1000000000000000000').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
```
- Loaded contracts on Foreign and then:
```
    ForeignBridge.methods.setHomeDailyLimit('5000000000000000000').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
    ForeignBridge.methods.setHomeMaxPerTx('1000000000000000000').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
```

6. The state of the bridge was the same.
```
    Home
    Remaining Daily POA Quota: 29,999,998.9 POA
    Locked POA in Bridge Contract: 1.00 POA

    Foreign
    Remaining Daily POA20 Quota: 14,999,999.9 POA20
    POA20 Tokens Amount: 1.00 POA20
```

7. Reset revised blocks by bridge. No double spend happened.
8. Used `Bridge UI` again, sent a few transfers. Worked as expected.

**Comment by akolotov** [comment]
Ok. Great! I will take a look at the gists one more time and will initiate the procedure of the contracts upgrade.

**Comment by patitonar** [comment]
@akolotov I made some corrections on the methods you suggested (replaced `requires` with `require` and `eth` with `ether`) and updated the gist to include them:
[Gist for HomeBridge](https://gist.github.com/patitonar/df585134034f5b6fbe01928fc6e00e36) 
[Gist for  ForeignBridge](https://gist.github.com/patitonar/0ff940d070ef5c4d3dbd70e44eeae361)

Then I followed the same steps as before, except at the moment of the upgrade of the contracts:

- On Home:

1. Deployed new version
```
eth deploy <PK> HomeBridge.bin --sokol // 0x56e5BE6733ebf726cAC18e473fcb73B83e1eD4AB
```
2. Upgrade contract version
```
HomeBridge.methods.upgradeFrom2To3().encodeABI() // '0xf91b853b'
EternalStorageProxy.methods.upgradeToAndCall('2', '0x56e5BE6733ebf726cAC18e473fcb73B83e1eD4AB', '0xf91b853b').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
```

- On Foreign:

1. Deployed new version
```
eth deploy <PK> ForeignBridge.bin --kovan // 0xdb18D3964a1A5dAA1Cad6cb0bfe3F9e35931c2Ea
```
2. Upgrade contract version
```
ForeignBridge.methods.upgradeFrom1To2().encodeABI() // '0xe1bb8044'
EternalStorageProxy.methods.upgradeToAndCall('2', '0xdb18D3964a1A5dAA1Cad6cb0bfe3F9e35931c2Ea', '0xe1bb8044').send({from: '0x3752F0a521b2E383693f658D6FC377953024C207', gas: 4000000})
```

The state of the bridge was the same after the upgrade, no double spend was produced by Rust Bridge and transactions performed using bridge-ui worked as expected.

**Comment by akolotov** [comment]
The upgraded version of contracts was applied to the ETH mainnet and POA Core:
* [Gist for HomeBridge](https://gist.github.com/patitonar/df585134034f5b6fbe01928fc6e00e36)
  * [Deployed contract](https://blockscout.com/poa/core/address/0x2a591e78dbf206248264500ac64ebee936b87714)
* [Gist for  ForeignBridge](https://gist.github.com/patitonar/0ff940d070ef5c4d3dbd70e44eeae361)
  * [Deployed contract](https://etherscan.io/address/0x2a591e78dbf206248264500ac64ebee936b87714)

**Pull request merged**